### DISCUSSION TITLE
Getting tpm chip to play with jetson nano devkit

### DISCUSSION THREAD

**Discussion by Achder** [OE4T/meta-tegra#936] [Q&A]
I'm desperately trying to get our STPM4raspi board to work with the jetson nano devkit.
I'm working according to the chip documentation: https://www.st.com/resource/en/application_note/an5714-integrating-the-st33tphf2xspi-and-st33tphf2xi2c-trusted-platform-modules-with-linux-stmicroelectronics.pdf

I have configured the pinmux to:

```
        spi1_mosi_pc0 {
		nvidia,pins = "spi1_mosi_pc0";
		nvidia,function = "spi1";
		nvidia,pull = <TEGRA_PIN_PULL_DOWN>;
		nvidia,tristate = <TEGRA_PIN_DISABLE>;
		nvidia,enable-input = <TEGRA_PIN_ENABLE>;
	};

	spi1_miso_pc1 {
		nvidia,pins = "spi1_miso_pc1";
		nvidia,function = "spi1";
		nvidia,pull = <TEGRA_PIN_PULL_DOWN>;
		nvidia,tristate = <TEGRA_PIN_DISABLE>;
		nvidia,enable-input = <TEGRA_PIN_ENABLE>;
	};

	spi1_sck_pc2 {
		nvidia,pins = "spi1_sck_pc2";
		nvidia,function = "spi1";
		nvidia,pull = <TEGRA_PIN_PULL_DOWN>;
		nvidia,tristate = <TEGRA_PIN_DISABLE>;
		nvidia,enable-input = <TEGRA_PIN_ENABLE>;
	};

	spi1_cs0_pc3 {
		nvidia,pins = "spi1_cs0_pc3";
		nvidia,function = "spi1";
		nvidia,pull = <TEGRA_PIN_PULL_UP>;
		nvidia,tristate = <TEGRA_PIN_DISABLE>;
		nvidia,enable-input = <TEGRA_PIN_ENABLE>;
	};

	spi1_cs1_pc4 {
		nvidia,pins = "spi1_cs1_pc4";
		nvidia,function = "spi1";
		nvidia,pull = <TEGRA_PIN_PULL_UP>;
		nvidia,tristate = <TEGRA_PIN_DISABLE>;
		nvidia,enable-input = <TEGRA_PIN_ENABLE>;
	};
```

And patched the device tree to this:

```
	hdr40_spi1: spi@7000d400 { /* SPI 1 to 40 pin header */
		status = "okay";

		st33htpm0: st33htpm@0 {
			status="okay";
			compatible = "st,st33htpm-spi";
			reg = <0>;
			#address-cells = <1>;
			#size-cells = <0>;
			spi-max-frequency = <25000000>;
			controller-data {
				nvidia,enable-hw-based-cs;
				nvidia,rx-clk-tap-delay = <7>;
			};
		};

		spi@1 {
			compatible = "tegra-spidev";
			reg = <0x1>;
			spi-max-frequency = <33000000>;
			controller-data {
				nvidia,enable-hw-based-cs;
				nvidia,rx-clk-tap-delay = <7>;
			};
		};
	};
```

But when I load the tpm_tis_spi driver nothing happens. 
Do you have any ideas? Is there something obviously wrong with my approach?
I've rebuild the image several times and flashed the board.

**Comment by madisongh** [comment]
There have been some issues in the past with SPI and with configuring the pins on the 40-pin header on some Nanos.  Take a look at
* #573 
* #694 
* #707 

There's also a [wiki page](https://github.com/OE4T/meta-tegra/wiki/SPI-support), which is brief and may not be fully up-to-date, but may help.

**Reply by Achder** [reply]
SPI seems to be working as expected. I think it's some device tree issue. 
I'm wondering if the controller-data is wrong.

**Comment by Achder** [comment]
I think this might be a timing issue. It works sometimes and sometimes not. 
The TPM timing specifications look like this:
![Time specifications](https://user-images.githubusercontent.com/11417131/158572496-bf6ec6e5-8e6a-498f-a070-60ca67493b1d.png)

Would I need to configure some of these controller-data properties:
- nvidia,tx-clk-tap-delay: Delays the clock going out to the external device
  with this tap value.
- nvidia,rx-clk-tap-delay : Delays the clock coming in from the external
  device with this tap value.
- nvidia,cs-setup-clk-count : CS setup timing parameter.
- nvidia,cs-hold-clk-count : CS hold timing parameter.
- nvidia,cs-inactive-cycles : Cycles to be inactive between two packets.
  CS Inactive between packets is disabled if value of this is zero.
- nvidia,clk-delay-between-packets: Clock delay  between packets by keeping
  CS active. For this, it is required to pass the Chip select
  as GPIO.

Any ideas?

**Reply by madisongh** [reply]
Have you tried just reducing the clock frequency?  From your device tree fragment, it looks like you're using 25MHz, but the data sheet extract you posted shows the chip handles 10MHz - 24MHz.

**Reply by Achder** [reply]
Ah sorry. Yeah, I'm switching those a lot. currently I'm on 15Mhz

**Reply by Achder** [reply]
Alright so this is strange. 

It works when:

- I let the device "warm up" - i.e. letting it run for 10 minutes
- after the warm up I need to reboot
- load driver and generate random number: works

It does not work when:

- I then unplug the device and let it "cool down" for 5 minutes 
- plug the device back in
- loading the driver fails

**Reply by dat-tran-matternet** [reply]
I see  this warm up issue with the UART as well. I also am having issue with SPI reliability. I posted something related here with no solution yet: https://forums.developer.nvidia.com/t/tx2-nx-spi-cs-issues/210936
